contracts/optimize/SlotHelper.sol

```diff
@@ -0,0 +1,158 @@
+
+    function putRaw(
+        mapping(uint256 => bytes32) storage slots,
+        bytes memory datas
+    ) internal returns (bytes32 mdata) {
+        // warn: if data as ptr to move,your should keep "data.length" by another ptr
+        uint256 len = datas.length;
+        mdata = encodeMetadata(datas);
+        if (len > FIRST_SLOT_DATA_SIZE) {
+            bytes32 value;
+            uint256 ptr;
+            assembly {
+                ptr := add(datas, add(0x20, FIRST_SLOT_DATA_SIZE))
+            }
+            for (
+                uint256 i = 0;
+                i < (len - FIRST_SLOT_DATA_SIZE + 32 - 1) / 32;
+                i++
+            ) {
+                assembly {
+                    value := mload(ptr)
+                }
+                ptr = ptr + 32;
+                slots[i] = value;
+            }
+        }
+    }
+
+    function encodeMetadata(bytes memory data)
+        internal
+        pure
+        returns (bytes32 medata)
+    {
+        uint256 datLen = data.length;
+        uint256 value;
+        assembly {
+            value := mload(add(data, 0x20))
+        }
+
+        datLen = datLen << LEN_OFFSET;
+        value = value >> SLOTDATA_RIGHT_SHIFT;
+
+        medata = bytes32(value | datLen);
+    }
+
+    function decodeMetadata(bytes32 mdata)
+        internal
+        pure
+        returns (uint256 len, bytes32 data)
+    {
+        len = decodeLen(mdata);
+        data = mdata << SLOTDATA_RIGHT_SHIFT;
+    }
+
+    function decodeMetadataToData(bytes32 mdata)
+        internal
+        pure
+        returns (uint256 len, bytes memory data)
+    {
+        len = decodeLen(mdata);
+        mdata = mdata << SLOTDATA_RIGHT_SHIFT;
+        data = new bytes(len);
+        assembly {
+            mstore(add(data, 0x20), mdata)
+        }
+    }
+
+    function getRaw(mapping(uint256 => bytes32) storage slots, bytes32 mdata)
+        internal
+        view
+        returns (bytes memory data)
+    {
+        uint256 datalen;
+        (datalen, data) = decodeMetadataToData(mdata);
+
+        if (datalen > FIRST_SLOT_DATA_SIZE) {
+            uint256 ptr = 0;
+            bytes32 value = 0;
+            assembly {
+                ptr := add(data, add(0x20, FIRST_SLOT_DATA_SIZE))
+            }
+            for (
+                uint256 i = 0;
+                i < (datalen - FIRST_SLOT_DATA_SIZE + 32 - 1) / 32;
+                i++
+            ) {
+                value = slots[i];
+                assembly {
+                    mstore(ptr, value)
+                }
+                ptr = ptr + 32;
+            }
+        }
+    }
+
+    function getRawAt(
+        mapping(uint256 => bytes32) storage slots,
+        bytes32 mdata,
+        uint256 memoryPtr
+    ) internal view returns (uint256 datalen, bool found) {
+        bytes32 datapart;
+        (datalen, datapart) = decodeMetadata(mdata);
+
+        uint256 dataPtr = memoryPtr;
+        assembly {
+            mstore(dataPtr, datapart)
```

Comment:
This requires memoryPtr:memoryPtr+32 is allocated for the data, which should be true for solidity.  But a few comments here can help explain.